utils/wasm-builder/src/prerequisites.rs

```diff
@@ -31,61 +33,101 @@ fn print_error_message(message: &str) -> String {
 
 /// Checks that all prerequisites are installed.
 ///
-/// # Returns
-/// Returns `None` if everything was found and `Some(ERR_MSG)` if something could not be found.
-pub fn check() -> Option<String> {
-	if !check_nightly_installed(){
-		return Some(print_error_message("Rust nightly not installed, please install it!"))
+/// Returns the versioned cargo command on success.
+pub(crate) fn check() -> Result<CargoCommandVersioned, String> {
+	let cargo_command = crate::get_nightly_cargo();
+
+	if !cargo_command.is_nightly() {
+		return Err(print_error_message("Rust nightly not installed, please install it!"))
 	}
 
-	check_wasm_toolchain_installed()
+	check_wasm_toolchain_installed(cargo_command)
 }
 
-fn check_nightly_installed() -> bool {
-	crate::get_nightly_cargo().is_nightly()
-}
+/// Create the project that will be used to check that the wasm toolchain is installed and to
+/// extract the rustc version.
+fn create_check_toolchain_project(project_dir: &Path) {
+	let lib_rs_file = project_dir.join("src/lib.rs");
+	let main_rs_file = project_dir.join("src/main.rs");
+	let build_rs_file = project_dir.join("build.rs");
+	let manifest_path = project_dir.join("Cargo.toml");
 
-fn check_wasm_toolchain_installed() -> Option<String> {
-	let temp = tempdir().expect("Creating temp dir does not fail; qed");
-	fs::create_dir_all(temp.path().join("src")).expect("Creating src dir does not fail; qed");
-
-	let test_file = temp.path().join("src/lib.rs");
-	let manifest_path = temp.path().join("Cargo.toml");
-
-	fs::write(&manifest_path,
+	write_file_if_changed(
+		&manifest_path,
 		r#"
 			[package]
 			name = "wasm-test"
 			version = "1.0.0"
 			edition = "2018"
+			build = "build.rs"
 
 			[lib]
 			name = "wasm_test"
 			crate-type = ["cdylib"]
 
 			[workspace]
 		"#,
-	).expect("Writing wasm-test manifest does not fail; qed");
-	fs::write(&test_file, "pub fn test() {}")
-		.expect("Writing to the test file does not fail; qed");
+	);
+	write_file_if_changed(lib_rs_file, "pub fn test() {}");
+	write_file_if_changed(
+		build_rs_file,
+		r#"
+			fn main() {
+				let rustc_cmd = std::env::var("RUSTC").ok().unwrap_or_else(|| "rustc".into());
+
+				let rustc_version = std::process::Command::new(rustc_cmd)
+					.arg("--version")
+					.output()
+					.ok()
+					.and_then(|o| String::from_utf8(o.stdout).ok());
+
+				println!(
+					"cargo:rustc-env=RUSTC_VERSION={}",
+					rustc_version.unwrap_or_else(|| "unknown rustc version".into()),
+				);
+			}
+		"#
+	);
+	write_file_if_changed(
+		main_rs_file,
+		r#"
+			fn main() {
+				println!("{}", env!("RUSTC_VERSION"));
+			}
+		"#
+	);
```

Comment:
That's rather a sophisticated way to ask the version? Out of curiosity, why we do it like that compared to just invoking `rustc --version` or something alike?

Reply:
The best solution would be `cargo rustc --version`, but that doesn't work. I want to know the rust version that is being used by the cargo command that I figured out is using nightly.

TLDR: There is no other way to get the rustc version I'm aware of.

Reply:
ah, thanks for the clarification. Mind leaving a comment in code in posterity?